### PR TITLE
 Use "frequently used emojis" for autocompletion in composer

### PR DESCRIPTION
**Pull request by grimhilt**
Fixes https://github.com/vector-im/element-web/issues/18978

Place "frequently used emojis" on top position in the composer autocomplete panel.

Signed-off-by: <grimhilt@users.noreply.github.com>

---
Here's what your changelog entry will look like:

## ✨ Features
 *  Use "frequently used emojis" for autocompletion in composer ([\#8998](https://github.com/matrix-org/matrix-react-sdk/pull/8998)). Fixes vector-im/element-web#18978. Contributed by @grimhilt.

### CODE REVIEW

File: src/autocomplete/EmojiProvider.tsx [hunk 1]
```diff
@@ -101,11 +105,9 @@ export default class EmojiProvider extends AutocompleteProvider {
 
         let completions = [];
         const { command, range } = this.getCurrentCommand(query, selection);
-
         if (command && command[0].length > 2) {
             const matchedString = command[0];
             completions = this.matcher.match(matchedString, limit);
-
```
Comment (SimonBrandner) [inline src/autocomplete/EmojiProvider.tsx:108]: Please don't change empty lines

File: src/autocomplete/EmojiProvider.tsx [hunk 2]
```diff
@@ -130,6 +132,15 @@ export default class EmojiProvider extends AutocompleteProvider {
             sorters.push(c => c._orderBy);
             completions = sortBy(uniq(completions), sorters);
 
+            completions = completions.slice(0, LIMIT);
+
+            //do a second sort to place emoji matching with frequently used one on top
```
Comment (SimonBrandner) [inline src/autocomplete/EmojiProvider.tsx:137]:
```suggestion
            // Do a second sort to place emoji matching with frequently used one on top
```

File: src/autocomplete/EmojiProvider.tsx [hunk 3]
```diff
@@ -130,6 +132,15 @@ export default class EmojiProvider extends AutocompleteProvider {
...
+            sorters.length = 0;
```
Comment (SimonBrandner) [inline src/autocomplete/EmojiProvider.tsx:138]:
```suggestion
            sorters = [];
```

This is more readable, IMO

**Review by SimonBrandner — COMMENTED** [review]
Thank you for working on this!

This is going to need a design review

File: src/autocomplete/EmojiProvider.tsx (hunk 1 above)
Comment (grimhilt) [reply src/autocomplete/EmojiProvider.tsx:108]: My bad, I though I check that

File: src/autocomplete/EmojiProvider.tsx (hunk 3 above)
Comment (grimhilt) [reply src/autocomplete/EmojiProvider.tsx:138]: I agree with you, but sorters is a constant

Comment (SimonBrandner) [reply src/autocomplete/EmojiProvider.tsx:138]: Then I would suggest making it into a `let` (or create a new variable)

**Review by SimonBrandner — CHANGES_REQUESTED** [review]
Looks sane codewise though I think it would be good to have tests for this, if possible

**Review by SimonBrandner — APPROVED** [review]
LGTM codewise!

**Comment by grimhilt** [comment]
Frequently used emojis are defined by the number of time there are been used, I didn't implement that, I used the same list as in the emoji picker under the category "Frequently used".

Current : 
image

Proposal :
image

The emojis present in the category "Frequently used" of the emoji picker (so the frequently used emojis) will be shown first for autocompletion.

**Review by niquewoodhouse — APPROVED** [review]

**Pull request merged**